### PR TITLE
Maths support

### PR DESCRIPTION
**Pull request by thosgood**
Linked to https://github.com/vector-im/riot-web/pull/10428

Added support for mathematics rendering using KaTeX (see https://github.com/vector-im/riot-web/issues/1945#issuecomment-514639318).

**Edit:** Now enabling maths mode replaces anything wrapped in dollar signs with the same thing wrapped in `<matrix-math>` (or `<matrix-math-display>`) tags in the `formatted_body`. Enabling KaTeX rendering (as an account option) then uses KaTeX to render anything between such tags.

This is an improvement over the previous commit, where the distinction between math mode and arbitrary `org.matrix.custom.html` content was non-existent. Now things really _can_ be changed to fit any proposals very easily (fallbacks, KaTeX vs. MathJax, delimiter choices, etc. etc.).

The only remaining 'bug' is that I haven't yet implemented per-room KaTeX rendering settings.

Signed-off-by: Tim Hosgood <timhosgood@gmail.com>

### CODE REVIEW

**Comment by thosgood** [comment]
closed while refactoring code to be more in keeping with proposals

File: res/img/math.svg [hunk 1]
```diff
@@ -0,0 +1 @@
....5857865 -2.1213203,0 c-.5857865,.5857864 -.5857865,1.5355339 0,2.1213203 l9.8994949,9.899495 c.3056756,.3056756 .7104567,.4518432 1.1109127,.438503 c.400456,.0133402 .8052371,-.1328274 1.1109127,-.438503 l9.899495,-9.899495 c.5857864,-.5857864 .5857864,-1.5355339 0,-2.1213203 c-.5857865,-.5857865 -1.535534,-.5857865 -2.1213204,0 l-7.4601551,7.4601551 v-16.5208153 c0,-.8284271 -.6715729,-1.5 -1.5,-1.5 s-1.5,.6715729 -1.5,1.5  z " visibility="visible" fill-rule="evenodd" fill="#d8d8d8"></path><rect stroke="#000000" x="5.875px" height="31.5px" y="7.5px" id="rect1" stroke-width="0" width="55.5px" fill="#d7d8d7" transform="" visibility="hidden"></rect><text stroke="none" style="outline-style:none;" id="text1" stroke-width="1px" x="48.625px" text-rendering="geometricPrecision" font-family="AmericanTypewriter-Bold" fill="#000000" font-size="27px" y="30.75px" transform="" text-anchor="middle" visibility="hidden">∑</text><text stroke="none" style="outline-style:none;" id="text2" stroke-width="1px" x="20.375px" text-rendering="geometricPrecision" font-family="AmericanTypewriter" y="37.125px" fill="#000000" font-size="40px" transform="" text-anchor="middle" visibility="hidden">$</text></svg>
```
Comment (Evidlo) [inline res/img/math.svg]: Is this right?  It looks like the markdown logo still.

Comment (thosgood) [reply res/img/math.svg]: I based it on the Markdown logo for some consistency:
Screenshot 2019-08-05 at 21 27 47

Comment (thosgood) [reply res/img/math.svg]: Screenshot 2019-08-05 at 21 28 40
(zoomed in)

Comment (Evidlo) [reply res/img/math.svg]: I mean [math.svg](https://github.com/matrix-org/matrix-react-sdk/blob/ffb596f0f73774a61c0e31575a5fa68258e96c1a/res/img/math.svg) specifically is the Markdown logo.  `math.png` is fine.

Comment (thosgood) [reply res/img/math.svg]: sorry, you're entirely right. I have no idea how `markdown.svg` was made, so just opted for a png instead, but I created a `math.svg` file (by copying the markdown one) just so the resource existed (for whenever somebody could make an svg)

Comment (thosgood) [reply res/img/math.svg]: just fixed this, thanks to `swedneck:permaweb.io`, so now the svg version is used instead (and the temporary png deleted)

**Comment by thosgood** [comment]
this has now entered testing at www.pigeon.digital , and the last commit fixes some problems reported by users

**Comment by thosgood** [comment]
animated gif of usage
Screen Recording 2019-08-22 at 16 32 07

**Comment by thosgood** [comment]
Matthew brought https://gitlab.com/gitlab-org/gitter/webapp/issues/1859 to my attention, but this was a case of the Gitter implementation of KaTeX being insecure. From what testing I've done, it seems like this PR doesn't have this security issue, and that HTML is properly escaped (since it doesn't use the autorender extension, for one). More details can be found at https://github.com/KaTeX/KaTeX/issues/1160.

It would obviously be good for a proper coder to have a look at this though, just to be sure.

**Comment by ara4n** [comment]
sorry that it's taken so long for us to get this properly on the radar - we're now taking a proper look. thank you for putting so much effort and patience into it!

**Comment by thosgood** [comment]
that's ok, just let me know if you have any questions or want me to do anything else! :)

**Comment by thosgood** [comment]
Following a bunch of discussion in `pigeon.digital:matrix.org`, this now follows the spec in https://github.com/matrix-org/matrix-doc/pull/2191 . There are still issues with per-room KaTeX settings, and fallback is only the code wrapped in `<code>` tags, but this is a first step towards what seems to be the consensus of how things should work.

**Comment by dkout** [comment]
Hey all, 
Are there any updates for this feature? Would be very useful to have! :)

**Comment by thosgood** [comment]
i haven't had any time to update this to work with the new message composer, and i'm not sure if much has been done since the last discussion in the room... i'd also love to finish this off but have really lost the swing of things, and can't remember where things were left off

**Comment by Moebiusgeometer** [comment]
Is anyone able to make this work as a test in the coming weeks as it was done previously above? I am a mathematics PhD at Paris-Saclay. We are all locked in at home now due to COVID-19. I would like to get students and teachers on Riot to continue courses, study groups, etc. Getting this running soon would be super helpful for many Universities worldwide in those days, even if buggy! I am unfortunately unable to do it for sure, but I promise to promote Riot (Matrix) to as many Universities and Professors I have contact to if someone manages to get it done.

**Comment by thosgood** [comment]
> Is anyone able to make this work as a test in the coming weeks as it was done previously above? I am a mathematics PhD at Paris-Saclay. We are all locked in at home now due to COVID-19. I would like to get students and teachers on Riot to continue courses, study groups, etc. Getting this running soon would be super helpful for many Universities worldwide in those days, even if buggy! I am unfortunately unable to do it for sure, but I promise to promote Riot (Matrix) to as many Universities and Professors I have contact to if someone manages to get it done.

There is a demo build running at https://pigeon.digital that should work fine. I believe it was recently used at a conference that wanted to do some remote interaction. Unless there's a massive rush of traffic, it should continue to work as a temporary solution, but I'm glad to hear that there's still interest in the mathematics community for this!

**Comment by eMPee584** [comment]
How much funding would this need to complete?
Our uni is going to put up a moderate matrix support budget..

**Comment by behrmann** [comment]
I haven't looked into this much further, then simply trying to rebase this on the current develop branch, since my JS-fu is basically non-existent, but this branch rebases almost cleanly, with only a trivial conflict in `src/HtmlUtils.js` and the need to bump the Katex version in `package.json` and have yarn fix the `yarn.lock`.

I'd really love to see this land, how can I help to make this happen?

**Comment by thosgood** [comment]
The last I heard was that this was going to be a Google Summer of Code project for somebody, because the team (very understandably) wanted any solution to not be simply for Riot, but for Matrix clients as a whole, which makes things quite a bit more difficult.

**Comment by behrmann** [comment]
@thosgood Thanks for the information! I just read the GSoC proposal and this PR is listed as a partial implementation. That is because it is only client side and there is no server endpoint that would generate a fallback (e.g. an image) or is there anything else missing? 

I guess most LaTeX users would be happy with the LaTeX code as fallback, as that is what (most of them) they are typing right now already.

**Comment by thosgood** [comment]
@behrmann No server fallback is the main thing missing here, yes. Not to mention that the message sent is just some custom type that mixes together a few different suggestions, whereas the "final" proposal might end up using extensible events or something like this (but this, again, seems to be a bit up in the air still).

I agree that, at least as a first goal, LaTeX implementation should really be aimed at the sort of people who use LaTeX often, i.e. not worrying too much if the fallback is just the code itself. Adding better fallbacks/better input methods (some WYSIWYG bar or something) seems like something that should come afterwards, to me.

**Comment by behrmann** [comment]
It seems like the Project unfortunately wasn't picked up by a student for the summer of code, or I am unable to find it on the [page](https://summerofcode.withgoogle.com/organizations/6060943798173696/). I'd still be happy if this would land, just as a stop gap until a more formal MSC is approved, as this would be quite a killer feature for the natural sciences.

**Comment by Adrien-Luxey** [comment]
The latest Element.io blog post references this issue, as a feature that will be implemented to ease the feelings of people disagreeing with the name change: https://element.io/blog/welcome-to-element/

So I guess LaTeX math support is on the roadmap after all!

**Comment by thosgood** [comment]
just as a small note, the current version of pigeon.digital uses the latest (pre-Element) build of Riot, i.e. with the new composer, and with all LaTeX rendering options hidden from the user (and all enabled by default), following user feedback

**Comment by behrmann** [comment]
@thosgood Sounds excellent. Do you know if the work is based in this PR or whether it is happening separately (I didn't see any other open PR). Would you be open to updating this PR (and the accompanying riot/element PR) to the version you seem to have in production? I'd love to give it a spin.

**Comment by thosgood** [comment]
@behrmann not too sure: all discussion (that I know of) is taking place in the pigeon.digital room. I could try updating this PR, but I think I'd be lost in git hell for a while in trying to do so... honestly, the changes that I had to make to the latest Riot release were so tiny: include the `katex` package, run the KaTeX renderer on any message with the `org.matrix.custom.html` format, and make sure that any message with a pair of `$` in it would be sent as this format. There is some extra playing around to be done with regexes to try to fix some problems that come from unmatched `$`s and how they interact with markdown things etc, but I didn't spend much time on fixing these, since these are the sorts of things that should be done on the same level as the markdown engine, and I didn't want to have to understand that yet 🙃

**Comment by behrmann** [comment]
@thosgood You can simply rename `thosgood:develop` via `git branch -m develop develop_v0` and then checkout the one with your current version as the new develop `git checkout -c develop <relevant branch>` und push it here via `git push --force-with-lease <name of your github remote> develop`. This should get you around the Git hell part. I think Github nowadays is also able to show different versions of a PR like Gitlab.

**Comment by polwel** [comment]
I just gave the `pigeon.digital` build a test drive. Nice work! Though editing messages seems to be broken.

**Comment by thosgood** [comment]
yeah, editing messages required a bit more finesse, and was never something that I gave much time to.

**Comment by thosgood** [comment]
as a warning, I think the `pigeon.digital` site will be unavailable soon, since the domain registration runs out, and I don't currently have the funds to renew it. when this happens though, i will definitely update this pull request with the latest version!

**Comment by polwel** [comment]
> yeah, editing messages required a bit more finesse, and was never something that I gave much time to.

Fair enough. Editing will be important though. I never seem to be able to get long LaTex expressions right on the first try.

**Comment by jonathancross** [comment]
@thosgood This has conflicts that need tp be resolved. Would be great to get a status update too. :-)

**Comment by thosgood** [comment]
this has been superseded by https://github.com/matrix-org/matrix-react-sdk/pull/5244 :-)

**Comment by ArniDagur** [comment]
I guess this should be closed as #5244 has been merged?

**Comment by jryans** [comment]
Yes, looks like we can close this now.

**Pull request closed without merging**